doc/design/python_interface.md

```diff
@@ -0,0 +1,99 @@
+# Python Interface Design
+
+## Variable as variable data type (not parameter)
+The class Variable is introduced to define the variables, which are not only parameters but also the input data and intermediate outputs.
+
+All the variables are defined in a variable scope, `Scope`, it will check duplicate variable names.
+
+The `Variable`'s definition is as follows
+```python
+class UniqueNameGenerator:
+    counter = 0
+
+    def __call__(self, prefix):
+        name = '%s-%d' % (prefix, counter)
+        counter += 1
+        return name
+
+
+unique_name_generator = UniqueNameGenerator()
+
+
+class Variable(object):
+    def __init__(self, shape, name=None, prefix=None, var_scope=None, op=None):
+        '''
+        var_scope is VarDesc's scope.
+        '''
+        if name is None:
+            if prefix is not None:
+                name = unique_name_generator(prefix)
+            else:
+                name = unique_name_generator("unknown")
+        self.name = name
+        self.var_scope = var_scope
+        self.op = op
```

Comment:
How to define a `varibale`'s `op`? If two operators write the same variable, which one is its `op`?

Reply:
the last one who write it? @reyoung @jacquesqiao